src/NServiceBus.Core/Extensibility/ContextBag.cs

```diff
@@ -143,8 +144,15 @@ internal void Merge(ContextBag context)
             }
         }
 
+        internal IBehavior[] Behaviors
+        {
+            get => behaviors ?? parentBag?.Behaviors;
```

Comment:
wouldn't this result in getters being chained all the time in further down stages till it ends up at the root Context that has the field actually set? Or is `behaviors` also set somewhere else besides the initial pipeline invocation?

Reply:
The setter sets the field directly on the current context. Each pipeline invocation for a specific context sets the array, see https://github.com/Particular/NServiceBus/pull/6237/files#diff-e945f887442cb120a1c4dfecb4621cab64bf9d2eb7089ab3edef39a57cdb81f7R45. If that setter is set it has precedence on the current behaviors, if not it walks up as it should be since the context does the same if I'm not mistaken. Have a look at the approval files all the pipelines seem to be built and executed correctly

Reply:
>  Each pipeline invocation for a specific context sets the array,

but that happens only once per full pipeline (e.g. incoming, outgoing, etc.), right?

> all the pipelines seem to be built and executed correctly

I'm not doubting the correctness, I was merely wondering whether this lookup (which will be quite a few stages deep in our pipelines) could be optimized away (e.g. by eagerly setting the reference on a new context when being created)

Reply:
Presently, setting the behaviors is done as part of the pipeline, which also keeps a reference to the behaviors. With what you propose, we would need to codify knowledge of pipeline spawning into the locations where contexts are created and the kind of reference to the pipeline behaviors from that point. Currently, I can't imagine a clean design how to do this. 

If you have some good ideas how to achieve this I'm all ears.

Reply:
couldn't you assign it in the ctor of the `ContextBag`? It would be `null` at the beginning until you explicitly set it as part of the `Invoke` method but from there one, creating new contexts for the other stages doesn't have to traverse all the way back to the parent that has the value set? Now I don't have any idea whether that would improve anything at all tbh (and it comes with some downside of impacting all other ContextBags used outside on the pipeline).

Reply:
Ah got it now. Very clever. 

![image](https://user-images.githubusercontent.com/174258/152245312-99360a2a-694a-4921-8782-d2b5524efa21.png)